spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,499 @@
+# These links could redirect us to potentially dangerous websites, and we don't want our users to click them!
+# To select the subset of shortened links in our dataset, we write a regex that checks for the commonly-used `.ly` extension.
+#
+# You'll notice that the slicing function is noisily defined — it doesn't represent the ground truth for all short links.
+# Instead, SFs are often heuristics to quickly measure performance over important subsets of the data.
+
+# %%
+import re
+from snorkel.slicing import slicing_function
+
+
+@slicing_function()
+def short_link(x):
+    """Returns whether text matches common pattern for shortened ".ly" links."""
+    return bool(re.search(r"\w+\.ly", x.text))
+
+
+sfs = [short_link]
+
+# %% [markdown]
+# For our $n$ examples and $k$ slices in each split, we apply the SF to our data to create an $n \times k$ matrix. (So far, $k=1$).
+
+# %%
+from snorkel.slicing import PandasSFApplier
+
+applier = PandasSFApplier(sfs)
+S_train = applier.apply(df_train)
+S_valid = applier.apply(df_valid)
+S_test = applier.apply(df_test)
+
+# %% [markdown]
+# ### Visualize slices
+
+# %% [markdown]
+# With a utility function, [`slice_dataframe`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/slicing/snorkel.slicing.slice_dataframe.html#snorkel.slicing.slice_dataframe), we can visualize examples belonging to this slice in a `pandas.DataFrame`.
+
+# %%
+from snorkel.slicing import slice_dataframe
+
+short_link_df = slice_dataframe(df_valid, short_link)
+short_link_df[["text", "label"]]
+
+# %% [markdown]
+# ## 2. Train a discriminative model
+#
+# To start, we'll initialize a discriminative model using our [`MultitaskClassifier`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.MultitaskClassifier.html#snorkel.classification.MultitaskClassifier).
+# We'll assume that you are familiar with Snorkel's multitask model — if not, we'd recommend you check out our [Multitask Tutorial](https://github.com/snorkel-team/snorkel-tutorials/blob/master/multitask/multitask_tutorial.ipynb).
+#
+# In this section, we're ignoring slice information for modeling purposes; slices are used solely for monitoring fine-grained performance.
+
+# %% [markdown]
+# ### Featurize Data
+#
+# First, we'll featurize the data—as you saw in the introductory Spam tutorial, we can extract simple bag of words features and store them as numpy arrays.
+
+# %%
+import torch
+from sklearn.feature_extraction.text import CountVectorizer
+
+vectorizer = CountVectorizer(ngram_range=(1, 1))
+
+
+def df_to_torch_features(vectorizer, df, fit_train=False):
+    words = [row.text for i, row in df.iterrows()]
+
+    if fit_train:
+        feats = vectorizer.fit_transform(words)
+    else:
+        feats = vectorizer.transform(words)
+    X = feats.todense()
+    Y = df["label"].values
+    return X, Y
```

Comment:
Generic code, orthogonal to the point of this nb- should be (a) put in util file, and (b) standardized across these tutorials as much as possible (where (b) might be separate PR)